### DISCUSSION TITLE
Tailwind workflow

### DISCUSSION THREAD

**Discussion by sebastian** [leptos-rs/leptos#1301] [General]
I have had a love a first sight experience with Leptos. The concepts, the way it is being built, and the community around it are all individually wonderful and make for a wonderful whole! I was really excited to see there being Tailwind support built right into cargo-leptos too!

I have spent years working with other framework, ranging from rails, phoenix (with elixir) and creating react apps in F# with Fable and frontends with Svelte. Having a quick turnaround time from making edits in my editor to seeing the changes reflected in the browser is something I have come to rely quite heavily on - particularly when working with Tailwind where I tend to make a lot of small incremental changes, seeing if the results are as I would like them to be.

I would love to learn what your workflow is when it comes to building and designing with Tailwind when using Leptos. 
In the app I am currently building it takes around 5 seconds before any change made to a component is reflected in the interface. This is quite frustrating, particularly so since I, still new to Leptos, often make mistakes and then stare at the browser in vain looking for my changes to appear, wondering if the change I made was invalid, I accidentally edited the wrong file, or whether I just accidentally broke the build. As a result I have come to rely on having a secondary sveltekit app on the side where I create and prototype my components, but that feels rather cumbersome and forces me to do a lot of copying and pasting, and spending time keeping two distinct apps in sync, only one of which is ever meant to see the light of day. Do you have a similar workflow, or do you see entirely different refresh times (i.e. is my setup broken? (I am currently using the latest version from the `main` branch on a modern MacBook Pro so am used to most things being fast))

The hot-reload experience in Dioxus is quite amazing actually. I imagine the underlying technology might be quite different, but I am hoping against hopes that something similar is in the future for Leptos too!

Looking forward to learning more!

**Comment by gbj** [comment]
One thing you might want to try right off the bat is `cargo leptos watch --hot-reload`, which is semi-experimental but very similar to the Dioxus approach. (i.e., works on most static HTML parts of a view, not things that need re-compilation via Rust) 

At the moment this works slightly better if you're running `npx tailwindcss` in a separate process instead of using the built-in cargo-leptos Tailwind support.

**Reply by sebastian** [reply]
That sounds wonderful!

Unfortunately I am not able to get this working. 
The help text for the `--hot-reload`-flag seems to indicate that one needs to run with nightly, so I have tried with:
- compile with tailwind directly, rather than through `cargo-leptos`
- use nightly rust (1.72.0 as of writing this)
- enable the nightly flags in the leptos libraries

All the same, the compile times are still as they previously were, even when editing components that do not have any dynamic state or behaviour.

I'll look into it a bit more and see if I can figure out what I have done wrong.

Either way, thank you for the very prompt reply, @gbj!

**Reply by gbj** [reply]
If the hot reloading is set  up and sending messages from cargo-leptos to the browser properly, you should see a little `[HOT RELOADING] Connected to server.` on initial load, then see messages sent through patching the DOM. I'm attaching a video showing this behavior, which should be easier to understand than a wall of text!

Let me know if
1) The various "hot reloading" messages aren't appearing in the browser console, or
2) Messages are appearing, but the DOM isn't being patched, or
3) Messages are appearing, and the DOM is being patched with new Tailwind classes, but the classes aren't being applied (because the new Tailwind CSS file is not being refreshed)

https://github.com/leptos-rs/leptos/assets/286622/f686848b-73ae-4ddf-8c51-f7d3b9c02b65

**Reply by sebastian** [reply]
Thank you for the video, good idea!

- ✅ I get hot-reloading messages
- ☠️ Patches aren't actually successfully applied
- ☠️ Firefox is complaining about some broken JS in the `main.js`, whereas neither Safari, nor Chrome seem to

So in the end it seems like the DOM patching fails, and hence I end up having to wait for a full refresh that happens as a result of the full slower compilation run.

Here is a video of what it looks like on my end:

https://github.com/leptos-rs/leptos/assets/1089/26853d82-c105-4532-afd5-fbba14e9dfc1

**Reply by sebastian** [reply]
And, for what it's worth, the JS error I get in Firefox is on the third line in this snippet:

```js
function ready(fn) {
    if (document.attachEvent ? document.readyState === "complete" : document.readyState !== "loading"){
        fn();
    } else {
        document.addEventListener('DOMContentLoaded', fn);
    }
}
```

**Reply by gbj** [reply]
The `main.js` thing is surely from a Firefox extension you have enabled or something — that code isn't anything from Leptos and I don't see it generated anywhere by wasm-bindgen, and if it's only in one browser I'd assume a browser extension. But it doesn't seem to be the issue here.

Looking at the video it looks like the line numbers in the patches that are being sent don't correspond to the line numbers in your actual file. I'd guess that's why it's not finding them in the DOM to update... Each invocation of the `view!` macro gets wrapped in comments indicating where in the source code it comes from, which is used to update it.

What is your `cargo leptos --version` and what version of `leptos` is in your Cargo.toml?

**Reply by sebastian** [reply]
Sorry, accidentally posted in [a new comment](https://github.com/leptos-rs/leptos/discussions/1301#discussioncomment-6397281) rather than this thread... But that aside: my original question, about workflow, you kindly answered already! And now it has turned more into a support/debugging issue... I am of course more than happy to continue debugging (after all I would love the hot-reload to work for me), but I just wanted to say that I really don't expect you to spend your time troubleshooting with me!

**Reply by sebastian** [reply]
Ok, from further testing, here is what seems to happen: hot reloading works _once_ after the server have started. Then I get an error from the second attempt at doing hot-reloading onwards. The error in question is this one:

<img width="667" alt="CleanShot 2023-07-09 at 22 20 47@2x" src="https://github.com/leptos-rs/leptos/assets/1089/d1a6cd60-2f81-483a-9567-afdd872a83a7">

**Reply by sebastian** [reply]
Please note that when starting fresh with a new leptos application, the hot reloading works! I'll recreate my app. It seems to be stuck in some old state.

**Reply by gbj** [reply]
Wow, that's awesome to hear! It's not a perfect system clearly, but I hope debugging it hasn't cost you more time than you'd save by it working 😂

Just kidding, I'm glad it's working okay now and feel free to open another discussion if it breaks and I'm happy to help debug.

**Comment by sebastian** [comment]
I was originally on:
- `cargo-leptos 0.1.11` (but I wonder if that is a bit deceiving. I installed the version from the `main`-branch yesterday, so it might be newer
- `leptos` is also from `main` as of yesterday (`source = "git+https://github.com/leptos-rs/leptos.git?branch=main#6b90e1babd425c9a324181c86e3fd1b942c9b10f"`).

I tried running the latest stable releases too (`0.4.2` of `leptos` and `cargo-leptos 0.1.11` of `cargo-leptos` but installed via crates.io), and had _some_ success with that! The hot reloading worked for a little while, but now it's all gone again, unfortunately.

https://github.com/leptos-rs/leptos/assets/1089/82458dec-0acb-4ce8-8a04-3bcfa5345fd7
